api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -66,7 +66,7 @@ message MatchPredicate {
 // HTTP headers match configuration.
 message HttpHeadersMatch {
   // HTTP headers to match.
-  repeated route.v3.HeaderMatcher headers = 1;
+  repeated type.matcher.v3.HeaderMatcher headers = 1;
```

Comment:
We can't change existing sites that use `route.v3.HeaderMatcher` without a breaking API change, see https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md. So, existing sites stay as-is, new sites get to switch to the new type. You could, if you want to do the extra step, deprecate `headers` here and add a new header matcher field pointing to the `type.matcher.v3.HeaderMatcher` proto.

Reply:
The purpose of introducing this new copy of HeaderMatcher was:

`MatchPredicate` depends on `route.v3.HeaderMatcher`.
`route.v3.HeaderMatcher` has a new route action for rewriting upstream response headers.
We need to use `MatchPredicate` in this rewriter to match on certain conditions before rewrite.

To break this cyclic dependency, we copied HeaderMatcher in a separate place.
Now we must change `MatchPredicate` to no longer depend on route.v3.HeaderMatcher. Let me know if you think we can work around this.

Reply:
As discussed offline, I think you'd need to copy `MatchPredicate` as well. Is there any chance that @snowp will land the new generalized matchers in time for the PR you are supporting?